v0 of templated service code

d00rman opened the pull request:
This is just v0. A lot of things are still missing, but let's start somewhere :)

d00rman committed:
Add the config files

d00rman committed:
Minor changes in package.json

d00rman committed:
Add v0 of the actual service template code

gwicke commented on routes/v1.js:
```diff
@@ -0,0 +1,48 @@
+ * Reads and returns the configuration file used to
+ * start the service. IMPORTANT NOTE: This is a sample
+ * only for demostrating promises, you should not
+ * expose this endpoint in real services.
```
Maybe it's better not to include unsafe examples. That config might contain private info like passwords.

Perhaps provide something like fibonacci instead?

d00rman replied:
Definitely. I put it there more as a place holder than the definitive version. Will address that soon (either in this or another PR)

gwicke commented:
I wonder if https://github.com/krakenjs/swaggerize-express could be useful for our purposes.

gwicke commented on app.js:
```diff
@@ -0,0 +1,41 @@
+
+
+var express = require('express');
+var fs = Promise.promisifyAll(require('fs'));
```
Better to use a local Promise implementation here. The global es6 Promise doesn't actually have `promisifyAll`.

Example:

``` javascript
var P = require('bluebird');
var fs = P.promisifyAll(require('fs'));
```

d00rman replied:
Actually, I was relying on `service-runner`'s globalisation of bluebird, but now that that's gone, I'll change the code accordingly.

d00rman committed:
Reflect newest changes in servisor/service-runner

Concretely:
* rename servisor to service-runner
* use our own global bluebird Promise object (BBPromise) now that
  service-runner does not expose it by default

d00rman committed:
Add a very short readme entry about getting started

d00rman committed:
Mini fixes in the MD of the readme file

d00rman committed:
Add some more info about running the code to the readme file

d00rman committed:
Add some options to the app

* disable the X-Powered-By header set by Express
* disable the ETag header, as APIs should do it themselves
* enable compression

d00rman committed:
Add the /_info/home endpoint

This endpoint exemplifies the usage of redirection in Express,in that
sends back a 301 response pointing to the service's home page URI as
specified in package.json

d00rman committed:
Some more app options: POST body parsers

d00rman committed:
Update the /v1 example

The /v1/conf example is gone. Instead, there is /v1/wiki(/uri) which
shows various techniques likely to be needed by service developers (see
the full description in the header of the example route)

gwicke commented on config.dev.yaml:
```diff
@@ -0,0 +1,43 @@
+# Statsd metrics reporter
+metrics:
+  type: txstatsd
+  host: localhost:8125
```
I moved the port to a separate port: property.

d00rman replied:
Ah ok. Will change. (Idem for the comment below)

gwicke commented on config.prod.yaml:
```diff
@@ -0,0 +1,43 @@
+# Statsd metrics reporter
+metrics:
+  type: txstatsd
+  host: localhost:8125
```
Same here (separate port attribute).

d00rman committed:
Separate metrics host and port

d00rman committed:
Minor: use direct properties on app for logger, metrics, etc., i.e. skip
app.locals

d00rman committed:
Use a more correct exmaple of getting info from the MW API

d00rman committed:
Switch to local bluebird Promise requires (get rid of global.BBPromise)

d00rman committed:
Minor syntax fixes

d00rman committed:
Set dev config logger level to 'trace'

d00rman committed:
Add testing capabilities to the template

This commit adds a simple test suite to the service template.
Concretely:
* add dependencies to mocha, jshint and friends
* add the test and coverage scripts
* add the helpers (assert.js, server.js, logStream.js)
* add tests for all of the current routes

d00rman commented:
Current status:
- the simple template infrastructure is in place
- there are some routes showing
  - service information retrieval
  - simple responses
  - dealing with named and optional URI parameters
  - the usage of bluebird and preq
  - test scripts and coverage

Still to do:
- DOM parsing
- logging && metrics
- proper error handling
- documentation

d00rman committed:
Add a static route, an HTML and some tests for the express app itself

d00rman committed:
Add an empty route template in routes/

d00rman committed:
Add some notes about testing and coverage to the readme file
